numpy/core/defchararray.py

```diff
@@ -51,21 +51,45 @@ def _use_unicode(*args):
     Helper function for determining the output type of some string
     operations.
 
-    For an operation on two ndarrays, if at least one is unicode, the
-    result should be unicode.
+    For an operation on two ndarrays, if at least one is a dtype with a
+    scalar that subclasses str or bytes, the result should be that
+    dtype. Otherwise, if at least one is unicode, the result should be
+    unicode.
     """
     for x in args:
-        if (isinstance(x, str) or
-                issubclass(numpy.asarray(x).dtype.type, unicode_)):
-            return unicode_
-    return string_
+        if isinstance(x, str):
+            return numpy.dtype(unicode_)
+        scalar_type = numpy.asarray(x).dtype.type
+        if isinstance(scalar_type, unicode_):
+            return numpy.dtype(unicode_)
+        elif (issubclass(scalar_type, (str, bytes))):
```

Comment:
Ack, this doesn't actually work, because it will trigger for bytes.  The test should be:
```
np.char.add(b"asdf", "jkl;")
```

This function is only used in `np.char.add` since you (nicely) split out `_is_unicode`.  The input is now always dtypes, but we should just remove it and replace it with `DType = type(np.result_type(arr1, arr2))`  (No direct access to `common_dtype` on the class currently.)

Now, maybe `result_type("S3", "U3")` should fail.  But if it should, then this path also should fail.

I am going to push these two changes in a bit and then probably put it in (prolly with that test).  Your DType will need to define the compatible "common dtype", but that seems fine.

Reply:
Nevermind, this is broken beyond repair anyway.  If bytes comes first, we somehow manage to concatenate it to bytes (casting the unicode to a bytes object first) and then store it, losing most of it (or getting garbage).
The reverse `np.char.add("asdf", b"jkl")` fails "correctly".

So I am not sure I will fix this.  A few notes:

1. `result_type` is probably more correct, but we would have to cast the input arrays, otherwise there is just no point.
2. `_get_num_chars` may work for your ASCII dtype, but only because its storage size is 1.  So this doesn't yet work _really_.

`_get_num_chars` needs fixing probably if you want things to really work.  But overall, but for `add()` all of this is moot:  Trying to get it fully right doesn't seem worthwhile. `__add__()` is the only truly binary function here and it may well be easier (annoying complexity wise) to make it a ufunc, then to find a clean way to generalize it to user dtypes...